File: Sources/ObjectsFactoryObjC/ObjectsFactory.m
//
// Flare
// Copyright © 2023 Space Code. All rights reserved.
//

#import "ObjectsFactory.h"

@implementation ObjectsFactory
+ (id)createInstance:(Class)class {
    return [[class alloc] init];
}

+(id)createInstance:(Class)class properties:(NSDictionary *)properties {
    if (![class isKindOfClass:[NSObject class]]) {
        return nil;
    }

    id object = [self createInstance:class];
    if ([object isKindOfClass:[NSObject class]]) {
        NSObject *nsObject = object;
        [properties enumerateKeysAndObjectsUsingBlock:^(id key, id valueObject, BOOL *stop) {
                [nsObject setValue:valueObject forKey :key];
        }];
    }

    return object;
}
@end

File: Sources/ObjectsFactory/Classes/ObjectsFactory.swift
//
// objects-factory
// Copyright © 2023 Space Code. All rights reserved.
//

import ObjectsFactoryObjC

public typealias ObjectsFactory = ObjectsFactoryObjC.ObjectsFactory

public extension ObjectsFactory {
    static func create<T: AnyObject>(_ classType: T.Type, properties: [String: Any] = [:]) throws -> T {
        if let instance = ObjectsFactory.createInstance(classType, properties: properties) as? T {
            return instance
        }
        throw NSError(
            domain: "ObjectsFactory",
            code: 1,
            userInfo: [NSLocalizedDescriptionKey: "Cannot create an instance of \(classType)"]
        )
    }
}
